components/sup/src/http_gateway.rs

```diff
@@ -139,20 +139,44 @@ impl AppState {
     }
 }
 
+#[derive(Debug, PartialEq, Eq)]
+pub enum ServerStartup {
+    NotStarted,
+    Started,
+    BindFailed,
+}
+
 pub struct Server;
 
 impl Server {
-    pub fn run(listen_addr: ListenAddr, gateway_state: Arc<RwLock<manager::GatewayState>>) {
+    pub fn run(
+        listen_addr: ListenAddr,
+        gateway_state: Arc<RwLock<manager::GatewayState>>,
+        control: Arc<(Mutex<ServerStartup>, Condvar)>,
+    ) {
         thread::spawn(move || {
+            let &(ref lock, ref cvar) = &*control;
             let sys = actix::System::new("sup-http-gateway");
 
-            server::new(move || {
+            let bind = server::new(move || {
                 let app_state = AppState::new(gateway_state.clone());
                 App::with_state(app_state).configure(routes)
-            }).bind(listen_addr.to_string())
-            .unwrap()
-            .start();
+            }).bind(listen_addr.to_string());
+
+            match bind {
+                Ok(b) => {
+                    let mut started = lock.lock().expect("Control mutex is poisoned");
+                    *started = ServerStartup::Started;
+                    b.start();
+                }
+                Err(e) => {
+                    error!("HTTP gateway failed to bind: {:?}", e);
+                    let mut started = lock.lock().expect("Control mutex is poisoned");
+                    *started = ServerStartup::BindFailed;
+                }
+            }
```

Comment:
Since this same line is identical in both arms of the `match`
```rust
                    let mut started = lock.lock().expect("Control mutex is poisoned");
```
I'd suggest moving it outside. And since we know we always want to update `started`, I'd suggest making that the return value of the `match`. That requires moving the `b.start()` call earlier, which I think makes sense because if that call were to block or panic, setting the mutex to `ServerStartup::Started` is incorrect. With all those changes:
```rust
            let mut started = lock.lock().expect("Control mutex is poisoned");
            *started = match bind {
                Ok(b) => {
                    b.start();
                    ServerStartup::Started
                }
                Err(e) => {
                    error!("HTTP gateway failed to bind: {:?}", e);
                    ServerStartup::BindFailed
                }
            };
```
If there was a panic, the mutex paired with the condvar would be poisoned, but since there's no way that the thread doing an unconditional `wait` to recognize this, the main thread would be blocked forever. But if we fix this, part of things, the `wait` can be changed to a `wait_timeout` (see other comment).

Reply:
I had purposely put the call to `lock.lock()` inside the match, even though it duplicated the line, because the lock won't get released until it goes out of scope, and I had some initial trouble getting this working because the mutex wasn't releasing the lock and the manager went into an infinite wait on `cvar`.  I'll definitely try your suggestion though, and see how it goes.

Reply:
Yeah, this change causes the infinite wait because the lock never gets released.

Reply:
Good check. I thought I had tried it out, but it absolutely makes sense that you'd need the mutex to be unlocked before `sys.run()` is called.